### PR TITLE
fix(deps): Upgrade sigs.k8s.io/controller-runtime to 0.17.4

### PR DESCRIPTION
**Pull request by sakoush**
**What this PR does / why we need it**:

This PR upgrades `controller-runtime` from `v0.11.2` to `v0.17.4` (~2 years difference). Note that there is a newer version of this package `v0.18.1` but it requires go `1.22` and we are currently at go `1.21`.

It required changes as the api has moved since then:
- `ctrl.NewManager`
- reconcilers (included dummy `context` parameter)
- watcher signatures
- secrets handlers return

**Which issue(s) this PR fixes**:

Fixes INFRA-680 (internal)

### TODO:

- [x] Integration test with and without clusterwide installation

### CODE REVIEW

File: operator/controllers/mlops/suite_test.go [hunk 1]
```diff
@@ -46,7 +45,7 @@ func TestAPIs(t *testing.T) {
 	RunSpecsWithDefaultAndCustomReporters(
 		t,
 		"Controller Suite",
-		[]Reporter{printer.NewlineReporter{}},
```
Comment (sakoush) [inline operator/controllers/mlops/suite_test.go:49]: this library doesnt exist anymore in `controller-runtime` and I dont think it is required for testing.

File: operator/main.go [hunk 2]
```diff
@@ -63,16 +66,16 @@ func main() {
 	logger := zap.New(zap.UseFlagOptions(&opts))
 	ctrl.SetLogger(logger)
 
-	watchNamespace := namespace
+	watchNamespaceConfig := map[string]cache.Config{namespace: {}}
```
Comment (sakoush) [inline operator/main.go:69]: description of the new api is [here](https://github.com/kubernetes-sigs/controller-runtime/blob/9bc967a711e9538cff4099645dd6df64c9ba8a74/pkg/cache/cache.go#L181-L192)

File: operator/controllers/mlops/seldonruntime_controller.go [hunk 3]
```diff
@@ -213,7 +212,7 @@ func (r *SeldonRuntimeReconciler) updateStatus(seldonRuntime *mlopsv1alpha1.Seld
 }
 
 // Find SeldonRuntimes that reference the changes SeldonConfig
-func (r *SeldonRuntimeReconciler) mapSeldonRuntimesFromSeldonConfig(obj client.Object) []reconcile.Request {
+func (r *SeldonRuntimeReconciler) mapSeldonRuntimesFromSeldonConfig(_ context.Context, obj client.Object) []reconcile.Request {
 	logger := log.FromContext(context.Background()).WithName("mapSeldonRuntimesFromSeldonConfig")
```
Comment (lc525) [inline operator/controllers/mlops/seldonruntime_controller.go:216]: would it make sense to pass the (currently unnamed) `Context` function argument here?

File: operator/controllers/mlops/server_controller.go [hunk 4]
```diff
@@ -184,7 +183,7 @@ func (r *ServerReconciler) updateStatus(server *mlopsv1alpha1.Server) error {
 }
 
 // Find Servers that need reconcilliation from a change to a given ServerConfig
-func (r *ServerReconciler) mapServerFromServerConfig(obj client.Object) []reconcile.Request {
+func (r *ServerReconciler) mapServerFromServerConfig(_ context.Context, obj client.Object) []reconcile.Request {
 	logger := log.FromContext(context.Background()).WithName("mapServerFromServerConfig")
```
Comment (lc525) [inline operator/controllers/mlops/server_controller.go:187]: similar context question as above

File: operator/controllers/mlops/suite_test.go (hunk 1 above)
Comment (lc525) [reply operator/controllers/mlops/suite_test.go:49]: appreciate the comment, helpful to know why the change was needed!

**Review by lc525 — APPROVED** [review]
lgtm, with just one comment/question

File: operator/controllers/mlops/seldonruntime_controller.go (hunk 3 above)
Comment (sakoush) [reply operator/controllers/mlops/seldonruntime_controller.go:216]: The caller doesnt have a context though, I will add a TODO in the code to look at it more deeply.

File: operator/controllers/mlops/server_controller.go (hunk 4 above)
Comment (sakoush) [reply operator/controllers/mlops/server_controller.go:188]: I have added a TODO in the code as per the answer in the previous review comment.

**Pull request merged**